collector/__init__.py

```diff
@@ -1,504 +1,507 @@
+        time_slot_length_days = self.time_slot_length.to(u.day).value
+        time_starts = helpers.round_minute(self.twilight_evening_12, up=True)
+        time_ends = helpers.round_minute(self.twilight_morning_12, up=True)
+
+        # n in an array with the number of time slots in a night.
+        n = ((time_ends.jd - time_starts.jd) / time_slot_length_days + 0.5).astype(int)
+
+        # Most general times: can convert into others.
+        # TODO: This does not work as currently done:
+        # starts = Time(['2000-01-01', '2000-01-02'])
+        # ends = Time(['2000-01-01 12:00', '2000-01-02 12:00'])
+        # n = [3, 4]
+        # ts = [np.linspace(start.jd, end.jd, i) for start, end, i in zip(starts, ends, n)]
+        # -> [array([2451544.5 , 2451544.75, 2451545]), array([2451545.5, 2451545.66666667, 2451545.83333333, 2451546])]
+        # Time(ts, format='jd') generates exceptions.
+        # We can do:
+        # self.times = [Time(t, format='jd') for t in ts]
+        # This will give us arrays per night of the times.
+        # We can then wrap all of this in a Time object to get an array, but it seems
+        # more advantageous to keep them separate so we can do nightly lookups?
+        # self.times = Time(self.times)
+        # -> <Time object: scale='utc' format='jd' value=[array ts above in 1D]>
+
+        # NOTE: We want this as a Python list because the entries will have different lengths!
+        self.times = [Time(np.linspace(start.jd, end.jd - time_slot_length_days, i), format='jd')
+                      for start, end, i in zip(time_starts, time_ends, n)]
+
+        # Pre-calculate the different times.
+        # We want these as Python lists because the entries will have different lengths.
+        # TODO: Thus, the np.vectorize approach will not work here, but we keep it here as documentation.
+        # self.utc_times = np.vectorize(lambda t: t.to_datetime('utc'))(self.times)
+        self.utc_times = [t.to_datetime('utc') for t in self.times]
+        self.local_times = [t.to_datetime(self.site.value.timezone) for t in self.times]
+        self.local_sidereal_times = [vskyutil.lpsidereal(t, self.site.value.location) for t in self.times]
+
+        def altazparang(pos: SkyCoord):
+            """
+            Common code to invoke vskyutil.altazparang for a number of positions and then properly
+            combine the results into three numpy arrays, representing, indexed by time:
+            1. altitude
+            2. azimuth
+            3. parallactic angle
+            """
+            alt, az, par_ang = zip(
+                *[vskyutil.altazparang(p.dec, lst - p.ra, self.site.value.location.lat)
+                  for p, lst in zip(pos, self.local_sidereal_times)]
+            )
+            return np.array(alt), np.array(az), np.array(par_ang)
+
+        # Calculate the parameters for the sun, joining together the positions.
+        self.sun_pos = SkyCoord([vskyutil.lpsun(t) for t in self.times])
```

Comment:
Yes... right now it is kind of a mish-mash of Python and numpy and astropy and is not standardized. I don't think we want to standardize on astropy since it is very computationally intensive, so I suspect that we will want to ultimately move to numpy arrays of float. Right now you can consider this more of a transitionary phase where I am moving from the old Collector and not worrying about performance yet, just to make sure the results are consistent with what we had before.

After that, I will move as much as possible to Python `datetime` / `timedelta` and `numpy` arrays of `float`, I think, in the interests of optimizing the code.

Reply:
One thing I do need to do before we do that is make sure that the `vskyutil` functions all work properly the way we want them to. I've had multiple versions of these with different functions that I didn't realize about until later. I think this library really needs to be standardized, have type hints included, and have warnings removed since there are many.

The function names are confusing - at least to me as a programmer (e.g. I don't know what `xair` does unless I go dig it up), and the code breaks a lot of Python best practices.